### ISSUE TITLE
Temporary input lag when autosaving a big file

### ISSUE THREAD

**Issue by Fef0** [xournalpp/xournalpp#2675]
**Affects versions :**
 - OS: Ubuntu Linux
 - Desktop environment: Gnome on X11
 - Version of Xournal++: latest commit
 - Installation method: nightly-ppa

**Describe the bug**
When you open a big .xopp file (let's say 50/60+ pages with only hand-written notes) and enable autosaving, when the autosaving routine starts CPU usage increases by a few %, the input get unresponsive and it's impossible to write, erase or move objects for a few seconds, at least on low-end hardware (My tablet has an Intel Pentium N4200 with 4GB of RAM).
The bigger the file, the longer the wait.

**To Reproduce**
1. Open a big .xopp file of hand-written notes, haven't tested with PDFs
2. Enable autosave
3. From time to time the input freezes and you can't write, delete or move objects.

**Expected behavior**
Is it possible to handle saving on a different thread in such a way that the user won't get stuck waiting while xournal++ saves a big file?
This might be extended for simple saving too, so users won't have to stop writing while xournal++ saves, since it's a slow process for big files.

**Comment by createyourpersonalaccount** [comment]
If you disable autosave and just save, is the behavior the same?

**Comment by Fef0** [comment]
@createyourpersonalaccount Once I disable autosave the input lag at intervals stops, but save is still pretty slow and it's a blocking action (as it is intended to be I think).

**Comment by Technius** [comment]
Thank for you finding the cause. I've been having this issue for months now and have never figured out what the cause was. I always worked around it by manually saving (ctrl+s), which unfreezes the input and rendering.

**Comment by Technius** [comment]
Do you know if this happens on `1.0.20`?

**Comment by Fef0** [comment]
@Technius I can't test right now, has the saving routine been updated recently? If not, there's a good chance that this issue might be a latent one that is noticeable only on huge files (we're talking about 50/60 pages of dense handwriting) and on slow hardware. What do you think?

**Comment by Dartypier** [comment]
Hi I think the issue is there also with 1.1.1, the redo/undo, delete and handwriting suffer from lag when autosaving large files (around 30 pages).

**Comment by Nei8t** [comment]
I would also like to bring this issue up, as it's still present on the latest version (1.2.1). I annotate through a long PDF textbook and always encounter this issue which really impacts the user experience. My autosave cycle is set to 3 minutes. Each time it kicks in, my CPU fans go full blast and all actions performed during this time are frozen for 15 seconds (on average). If I try to perform actions during the autosave process, such as delete, redo, paste etc, it will only take action after the autosave process finishes (and all these actions would be performed at once). 
So basically every 3 minutes, I have to stop all of my work for 15 seconds for the autosave to finish. I am not sure what is the cause of this, but surely something can be done to mitigate it. This is also not a hardware problem as my system is pretty modern (9900k & 16GB RAM).

**Comment by createyourpersonalaccount** [comment]
@Nei8t Is it related to a particular setup or does it happen with all documents with autosave? Does it happen on the first autosave (and each after) or after a while?

**Comment by Nei8t** [comment]
> @Nei8t Is it related to a particular setup or does it happen with all documents with autosave? Does it happen on the first autosave (and each after) or after a while?

I have observed this behavior mostly affecting documents with a lot of pages. My PDF textbooks often consist of 300+ pages (I don't annotate all of them, 90% are left empty), and it seems to contribute a lot to the file size as well, I think. The file size of my documents ranges from 30MB to 60MB which I am not sure is ideal. Opening the xopp file without the attached PDF made no change to this behavior as well.

I am attaching a video demonstrating the issue. The autosave cycle kicks in at 0:07, in addition I've pressed a Ctrl-Z a few times at ~0:18. 
The changes kick in all at once after the autosave cycle finishes. 

https://github.com/xournalpp/xournalpp/assets/138250007/1d8aa3f5-9a3e-415c-8b51-73b701786d22

**Comment by createyourpersonalaccount** [comment]
@Nei8t Does it happen on the first autosave (and each after) or after a while? If you could also share a .xopp file or instructions to reproduce the bug, that would help. I'll try to investigate it, but I'll wait for your response first.

**Comment by Nei8t** [comment]
> @Nei8t Does it happen on the first autosave (and each after) or after a while? If you could also share a .xopp file or instructions to reproduce the bug, that would help. I'll try to investigate it, but I'll wait for your response first.

Apologies for the late reply, I was actually not able to reproduce the issue. It happens on the first autosave and each after. It surely relates to the file size, as this only happens with documents of size 25MB+ and it gets worse the larger the file size is. Though I believe my xopp documents may get unreasonably large. I have attached one file I have where I encountered the issue, though it gets worse with other larger documents I have (github has a 25MB upload limit). Alternatively, if you can manage to create a xopp document that is more than 30MB, that is enough to reproduce the issue.
[pythonbasics.zip](https://github.com/xournalpp/xournalpp/files/12476069/pythonbasics.zip)

**Comment by createyourpersonalaccount** [comment]
@Nei8t thanks I was able to replicate it. I will make an attempt to figure out why it's doing this.

**Comment by createyourpersonalaccount** [comment]
From a quick investigation, the line that takes time to save is https://github.com/xournalpp/xournalpp/blob/e9ec033031db7a3a1cfca40af93353d161a01f3d/src/core/control/jobs/AutosaveJob.cpp#L49

Maybe this line could run in its own thread. I'll try to write a patch when I'm less tired.

**Comment by createyourpersonalaccount** [comment]
Does the Scheduler class not use threads? @Technius @Dartypier 
I'm a bit surprised by this, are the scheduled jobs all ran on a single thread?

**Comment by bhennion** [comment]
This should have been mitigated by #7230: now navigation no longer freezes while autosaving. Writing to the document will still have to wait until the document has been saved though.
